### PR TITLE
Langmuir turbulence parameterization improvements

### PR DESCRIPTION
**Pull request by alperaltuntas**
# Pull Request Summary
 

This PR cleans up the calculation of Stokes drift and other diagnostics variables in WW3 for Langmuir turbulence parameterization in CESM, and removes the CESMCOUPLED ifdefs.

## Description
* Provide a detailed description of what this PR does.
  * Add Langmuir turbulence parameterization
* What bug does it fix, or what feature does it add?
  * Feature - allows enabling calculation of Stokes drift diagnostic quantities.
* Is a change of answers expected from this PR?
  * No.  Though the mod_defs will need to be updated.  

Please also include the following information: 
* Add any suggestions for a reviewer 
* Mention any labels that should be added:  
  * _enhancement_, _new feature_
* Are answer changes expected from this PR? Please describe the changes and the reason why in addition to which of the following labels would apply: 
  * Clean up the calculations of the surface Stokes drift (`USSX` and `USSY`) and surface layer averaged Stokes drift (`USSHX` and `USSHY`). The calculation of surface layer averaged Stokes drift and the tail contributions to both versions of Stokes drift are activated when `LMPENABLED` flag in mod_def file is set to True. Tail contribution is added in langmuir mixing computations if `SDTAIL` flag is set to True. HSLMODE switch determines how HSL (surface layer depth) is specified. HSLMODE=0 means testing (where HSL is set to 10m everywhere), HSLMODE=1 means HSL is received from the nuopc cap.
  * Change the input field of mixed layer depth `HML` to a more general surface layer depth `HSL`. This depth is used to calculate the surface layer averaged Stokes drift. When used in CESM, the surface layer depth is defined as 1/5 of the mixed layer depth. Now this 1/5 factor is applied to the mixed layer depth when importing the fields in the cap.
  * Clean up the diagnostic variables for Langmuir turbulence parameterization. Some variables are intermediate quantities and are not used in the parameterization. These variables are now deleted. The key variable diagnosed in WW3 is the surface layer averaged Stokes drift (`USSHX` and `USSHY`). The Langmuir enhancement factor (`LAMULT`) and the surface layer averaged Langmuir number (`LASL`) are now calculated when they are exported in the cap. `LASL` is needed for the Langmuir induced entrainment parameterization in CVMix but was not there in the cap code. So it is now added following the way `LAMULT` is defined.  
  * This PR only changes the diagnostic variables in WW3 and shouldn't change the answers of WW3. 

### Issue(s) addressed

* https://github.com/NOAA-EMC/WW3/issues/669
* https://github.com/NOAA-EMC/WW3/issues/670
* https://github.com/NOAA-EMC/WW3/issues/671 

### Commit Message

Clean up and refine Langmuir turbulence parameterization computations.
Authors: @qingli411, @alperaltuntas

### Check list  

- [x] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [x] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [x] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [x] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested?
  * Testing was done in both standalone mode and coupled within UFS, documentation: escomp/ww3/pull/19 
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)
* Have the matrix regression tests been run (if yes, please note HPC and compiler)?
* Please indicate the expected changes in the regression test output, (Note the [list](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-develop#4-look-at-results) of known non-identical tests.)
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):

### CODE REVIEW

**Comment by alperaltuntas** [comment]
NOTE: this PR is currently a draft for an initial review from @MatthewMasarik-NOAA and/or @JessicaMeixner-NOAA. I am also mentioning @mvertens and @DeniseWorthen, who may be interested in reviewing this early version.

**Comment by MatthewMasarik-NOAA** [comment]
Awesome, @alperaltuntas, it's great to see this work!  I have the UFS coupled tests running with it now.  I'll post an update tomorrow.

**Comment by MatthewMasarik-NOAA** [comment]
@alperaltuntas the runs I submitted yesterday had failures due to the mod_def's.  This I should have known since the PR modifies `w3gridmd.F90`, I needed to generate new mod_def's before running.  I will be out next week, but will return and resolve this the following week.  I discussed this PR with @JessicaMeixner-NOAA yesterday, and we will leave this in the Draft state while I'm away.

**Review by JessicaMeixner-NOAA — COMMENTED** [review]
@alperaltuntas --- First and foremost - THANK YOU!   These updates so far look good  and I'm planning on taking an even closer look when I return from a conference at the end of July. 

A couple of things that would be helpful in the meantime: 
* Updates to model/inp/* and model/nml/* as appropriate to document new additions 
* Add regression test for new features/output parameters  
* Any updates that should go into the manual (Alternatively, if you'd create an issue with an assigned person + what needs to be document that will be sufficient, we'll just want to make sure we get it before the next public release).

**Comment by MatthewMasarik-NOAA** [comment]
> @alperaltuntas --- First and foremost - THANK YOU! These updates so far look good and I'm planning on taking an even closer look when I return from a conference at the end of July.
> 
> A couple of things that would be helpful in the meantime:
> 
>     * Updates to model/inp/* and model/nml/* as appropriate to document new additions
> 
>     * Add regression test for new features/output parameters
> 
>     * Any updates that should go into the manual (Alternatively, if you'd create an issue with an assigned person + what needs to be document that will be sufficient, we'll just want to make sure we get it before the next public release).

Thank you @JessicaMeixner-NOAA for this initial look and comments.  @alperaltuntas I'll start to add in the items Jessica requested, so there's nothing additional needed at this point.  I'll work first to get the namelist and regression test implemented.  After that maybe we can touch base and together try to tackle some supporting documentation.

**Review by mvertens — CHANGES_REQUESTED** [review]
I've suggest a load balancing exercise and some minor changes along with questions. 
Frankly - I must not have submitted the suggestion - so please ignore this.

**Comment by alperaltuntas** [comment]
> I've suggest a load balancing exercise and some minor changes along with questions.

@mvertens Could you clarify? I am not sure what you mean by load balancing exercise, and which minor changes and questions you are referring to.

**Comment by DeniseWorthen** [comment]
@alperaltuntas I apologize for not responding earlier. I didn't see any issues w/ your changes in terms of the mesh cap. 

At some point, if this feature is used for the unstructured mesh, I think there will need to be a few changes to allow the new fields (usshx,usshy) to be written as gridded output. But I wouldn't view that as a requirement for this PR.

@MatthewMasarik-NOAA I assume you've run (or will run) the UFS RTs with these changes prior to merge?

**Comment by MatthewMasarik-NOAA** [comment]
> @MatthewMasarik-NOAA I assume you've run (or will run) the UFS RTs with these changes prior to merge?

@DeniseWorthen, yes, definitely.  I'm going to make a few edits for the added regtest, which I plan to complete this week.  Then @JessicaMeixner-NOAA and will review including running the UFS RTs.

**Review by mvertens — APPROVED** [review]

**Comment by DeniseWorthen** [comment]
@MatthewMasarik-NOAA Thanks. I will be able to approve once all the UFS are shown to pass.

**Comment by MatthewMasarik-NOAA** [comment]
@alperaltuntas  Glad to see the polishing updates were merged in.  Jessica will start her review on Monday.

File: regtests/ww3_tp2.22/input/ww3_grid.inp
Comment (JessicaMeixner-NOAA) [inline regtests/ww3_tp2.22/input/ww3_grid.inp]: Is there any reason to keep this here commented out, or should we go ahead and clean up this line and line 12?

File: regtests/ww3_tp2.22/input/namelists_2-D.nml
Comment (JessicaMeixner-NOAA) [inline regtests/ww3_tp2.22/input/namelists_2-D.nml]: In this and the next name list file I don't see: 
 &LMPN LMPENABLED = T, SDTAIL = T, HSLMODE = 0 /

File: regtests/ww3_tp2.22/input/gx_outf.inp
Comment (JessicaMeixner-NOAA) [inline regtests/ww3_tp2.22/input/gx_outf.inp]: Are you actually looking at these outputs? If so great! Otherwise, perhaps let's not add this and the *.gs file here.

File: regtests/ww3_tp2.22/input/namelists_2-D.nml
Comment (JessicaMeixner-NOAA) [reply regtests/ww3_tp2.22/input/namelists_2-D.nml]: Also I don't see a ww3_grid.nml file so I don't think this file is ever even used.   Since there are added values here, it might be worth making a test of nml since that's the way we're going.  At minimum an issue to test that should be made.

File: regtests/ww3_tp2.22/input/ww3_ounf.inp
Comment (JessicaMeixner-NOAA) [inline regtests/ww3_tp2.22/input/ww3_ounf.inp]: The output is to add things to stoke's drift, but I don't see us outputting stoke's drift here, perhaps this should be added to properly test the new features?

File: regtests/ww3_tp2.22/input/ww3_multi.inp
Comment (JessicaMeixner-NOAA) [inline regtests/ww3_tp2.22/input/ww3_multi.inp]: Add Stoke's drift variables to ensure the new test is properly testing and showing capabilities of new features.

Comment (JessicaMeixner-NOAA) [reply regtests/ww3_tp2.22/input/ww3_multi.inp]: I don't see a multi_grid test added to matrix.base, so perhaps removing the multi_related files as these aren't needed?

File: regtests/ww3_tp2.22/input/ww3_shel.inp
Comment (JessicaMeixner-NOAA) [inline regtests/ww3_tp2.22/input/ww3_shel.inp]: Add Stoke's Drift and other relevant output to this test to properly show new features capability and test to ensure things are working properly and that we do not break this feature in the future.

File: model/src/w3adatmd.F90
Comment (JessicaMeixner-NOAA) [inline model/src/w3adatmd.F90:2164]: This is adding a new output variable, but I see no mention of it in the description of the PR or in the model/inp or nml variables.

**Review by JessicaMeixner-NOAA — CHANGES_REQUESTED** [review]
First and foremost, thank you for all of these updates! This is really exciting and I look forward to getting this merged into dev/ufs-weather-model and to the develop branch.   From what I can tell looking at the code, there's an additional output variable but it's not mentioned in the manual or in any of the model/inp or model/nml files.

Thanks for adding this regression test which will help us ensure these functions stay working as expected.  This test adds a tail calculation to Stokes drift, but Stokes drift is not used in the added test, nor is the output parameter.  Before we merge this in, we should make sure this test is cleaned up and properly testing the added features so that we can make sure they continue to work as expect.  Thanks again for all of these exciting science updates as well as the huge amounts of code clean-up.  We look forward to merging this very soon!

**Comment by MatthewMasarik-NOAA** [comment]
@JessicaMeixner-NOAA thanks for the review.  Looks like there's a couple items to clean up still.  

@alperaltuntas these remaining items -- cleaning the unnecessary files and adding the outputs might be easiest for me.  I can work to address these.  For a timeline lets see what next week brings w.r.t. a government shutdown.

File: regtests/ww3_tp2.22/input/namelists_2-D.nml
Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.22/input/namelists_2-D.nml]: @JessicaMeixner-NOAA @alperaltuntas  
Regarding inp vs. nml files, since we have not switched over to nml yet, and not all the nml have been tested, I removed those scripts.  My idea was to keep the PR narrowly focused on the 3 issues it resolves and not branch out to testing the nml files.  I will create an issue for the nml file functionality though.

Comment (JessicaMeixner-NOAA) [reply regtests/ww3_tp2.22/input/namelists_2-D.nml]: sounds good

File: regtests/ww3_tp2.22/input/ww3_grid.inp
Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.22/input/ww3_grid.inp]: Addressed by: https://github.com/ESCOMP/WW3/pull/15/commits/5b1a73afac53a1cb2d5494caa14ef16039577f98

File: regtests/ww3_tp2.22/input/namelists_2-D.nml
Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.22/input/namelists_2-D.nml]: An issue (#1111) has been added to address this comment.

File: model/inp/ww3_grid.inp
Comment (JessicaMeixner-NOAA) [inline model/inp/ww3_grid.inp]: Are USS and USSH namlists associated with LMPN?  I don't see these being used in the regression test as a ww3_grid.inp namelist, but I do see this described in the model/inp/ww3_shel.inp and being used as output variables there.  I think lines 449-453 should possibly be deleted.

File: regtests/bin/matrix_cmake_ncep
Comment (JessicaMeixner-NOAA) [inline regtests/bin/matrix_cmake_ncep:206]: should this get added to the other matrix_cmake_*?

File: regtests/ww3_tp2.22/input/ww3_ounf.inp
Comment (JessicaMeixner-NOAA) [inline regtests/ww3_tp2.22/input/ww3_ounf.inp:16]: Any reason this doesn't match the output for ww3_shel:   HS EF T01 DIR SPR USS USSH 

It's fairly typical of our regtests, but since this one is new, maybe we can update this.

**Comment by JessicaMeixner-NOAA** [comment]
I have started to review and run tests.  I hope to have test results and review comments posted in the next day or two.

**Comment by MatthewMasarik-NOAA** [comment]
> I have started to review and run tests. I hope to have test results and review comments posted in the next day or two.

@JessicaMeixner-NOAA thanks for the update!  Fyi, @alperaltuntas

File: regtests/ww3_tp2.22/input/ww3_ounf.inp
Comment (JessicaMeixner-NOAA) [inline regtests/ww3_tp2.22/input/ww3_ounf.inp]: Since this is a new test, should we use NetCDF v4 or is there a particular reason to use v3?

**Review by JessicaMeixner-NOAA — COMMENTED** [review]
@alperaltuntas Thanks again for this PR.  My testing is ongoing due to some hiccups on my end.  I do have a few minor comments and wanted to make sure I asked those while tests are continuing.

**Comment by MatthewMasarik-NOAA** [comment]
> @alperaltuntas Thanks again for this PR. My testing is ongoing due to some hiccups on my end. I do have a few minor comments and wanted to make sure I asked those while tests are continuing.

@alperaltuntas @JessicaMeixner-NOAA, I can address these comments. I'll fill in replies after the repo meeting this morning.

**Comment by JessicaMeixner-NOAA** [comment]
One other minor comment, we need to get an issue created to add to the manual the new output parameter USSH.

**Comment by MatthewMasarik-NOAA** [comment]
@alperaltuntas, we're working on the testing, could you sync this branch to the NOAA-EMC/WW3: dev/ufs-weather-model branch when you have a moment?

File: regtests/bin/matrix_cmake_ncep
Comment (MatthewMasarik-NOAA) [reply regtests/bin/matrix_cmake_ncep:206]: Yes!  Good catch, this was an oversight.  I'll make a PR to this one with changes (fyi, @alperaltuntas)

File: regtests/ww3_tp2.22/input/ww3_ounf.inp
Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.22/input/ww3_ounf.inp:16]: Hmm, I don't know how I got the mis-match.  I'll adjust it to contain `HS EF T01 DIR SPR USS USSH`.

File: regtests/ww3_tp2.22/input/ww3_ounf.inp
Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.22/input/ww3_ounf.inp]: I used the `tp2.1` `ww3_grid.inp` file which had that setting.  I'll update that.  I also just realized I documented that `tp2.2` was used as a template for  `tp2.22`, though it was actually `tp2.1`, and `tp2.2` does not exist. I'll update any documentation that has that.

File: model/inp/ww3_grid.inp
Comment (MatthewMasarik-NOAA) [reply model/inp/ww3_grid.inp]: `USS` and `USSH` are not namelists, only output variables.  They are associated with the `LMPN` namelist in that they won't be available unless `LMPN` is turned on.  I listed both the input parameters and output parameters associated with `LMPN`.  I can remove `USS` and `USSH` if it's confusing or doesn't belong.  Please let me know and I can add it to the small fix PR I'll make.

Comment (JessicaMeixner-NOAA) [reply model/inp/ww3_grid.inp]: Surface Stokes drift is available without LMPN namelist. It's my understanding that SDTAIL=T adds a tail contribution to USS/USSH.   I find it confusing to add output parameters within the namelist description section and therefore recommend removing this here. The manual would be a good spot for any additional information that needs to be added potentially.

Comment (MatthewMasarik-NOAA) [reply model/inp/ww3_grid.inp]: Okay, I'll remove the output parameters from the namelist section to avoid confusion. 

I had been under the impression that the calculations for USS/USSH are turned on when `LMPENABLED` is set to `T`.  Looking back over it, I think think I misunderstood, and it is just the case for USSH.

File: regtests/ww3_tp2.22/input/ww3_ounf.inp
Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.22/input/ww3_ounf.inp]: fyi, `tp2.2` does exist, my mistake.

**Comment by MatthewMasarik-NOAA** [comment]
@JessicaMeixner-NOAA a PR to #1034, https://github.com/ESCOMP/WW3/pull/16, should resolve all the comments above.

**Comment by MatthewMasarik-NOAA** [comment]
Ready for review again.

File: regtests/ww3_tp2.22/input/ww3_ounf.inp
Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.22/input/ww3_ounf.inp:16]: @JessicaMeixner-NOAA I looked into what was going on here.  The mismatch with ww3_shel was already present in `ww3_tp2.2` and we just used that regtest as a base to demonstrate the `USS` and `USSH`.  When EF was added into `ww3_outf.inp` to match `ww3_shel.inp`, ww3_outf crashes because the parameter `EF` has not been implemented in `ww3_outf.F90`.  It would take time to implement and test if this is a direction we want to go.  To me this seems outside of the scope of this PR.  My recommendation would be to just revert back to what it was before.

Comment (JessicaMeixner-NOAA) [reply regtests/ww3_tp2.22/input/ww3_ounf.inp:16]: Adding EF to outf is out of the scope of this PR.  We should add an issue about this just for our own record keeping. 
  Does EF work in ounf?  If so, we could just keep it there, remove it from here  assuming it's critical for testing in this PR.  An alternative could be that EF is removed from shel, outf and ounf for consistency.

Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.22/input/ww3_ounf.inp:16]: It doesn't look like ounf wrote EF.  Sure, I can make remove it from shel and ounf

File: regtests/ww3_tp2.22/input/ww3_ounf.inp
Comment (JessicaMeixner-NOAA) [reply regtests/ww3_tp2.22/input/ww3_ounf.inp]: Since this is a new test and we're making other changes, let's go ahead and move to NetCDF v4 instead of 3.

Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.22/input/ww3_ounf.inp]: This change has already been made.

Comment (JessicaMeixner-NOAA) [reply regtests/ww3_tp2.22/input/ww3_ounf.inp]: Apologies, it didn't show up correctly when I looked a minute ago.  I see that it's changed now.  Thank you!!!

**Comment by MatthewMasarik-NOAA** [comment]
**Update**

@JessicaMeixner-NOAA @alperaltuntas  -  happy new year!  I wanted to give a status update after a bit of quiet during the holidays.  

The PR that addressed comments above regarding parameter `EF` (escomp/ww3/pull/16) broke the added test, `ww3_tp2.22`, because output of `EF` is not implemented in `ww3_outf.F90`.  I have a new branch that removes this parameter from the relevant `*.inp` files.  This addresses the comments above and resolves the crash.  

For testing in the UFS-WM, new `mod_def.ww3` need to be re-generated first, then RT's can be run. Once our R&D machine returns later today, here's how I'm planning to proceed.

1. Generate mod_defs and test in UFS-WM (by Fri, 1/12).
2. Submit new PR to ESCOMP (by Fri, 1/12).
3. Once the PR in 2. is merged in by @alperaltuntas, then
4. The final review can be started by @JessicaMeixner-NOAA

**Comment by MatthewMasarik-NOAA** [comment]
**update**

@alperaltuntas @JessicaMeixner-NOAA 

I apologize for not keeping folks on the forum up to date, there as been development and testing offline.  I'm finishing the testing of a PR branch to address the previous comments.  The standalone has been completed.  Once I've finished the coupled testing I'll submit that to this PR.  At that point this PR will be cleared to move forward.

**Review by JessicaMeixner-NOAA — COMMENTED** [review]
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/14558841/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/14558842/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/14558843/matrixDiff.txt)

Changes are in the mod_defs and out_grd as expected.  Plan to do one last test of ensuring the new regtests reproduce itself, but this PR should be good to have a ufs-weather-model PR submitted.

**Review by JessicaMeixner-NOAA — APPROVED** [review]
Thank you @alperaltuntas and @MatthewMasarik-NOAA for getting this in shape for eventual merge back to the develop branch. 

@ukmo-ccbunney @mickaelaccensi @thesser1 -  These updates will eventually be coming back to develop.  If there's anything you see that needs to be immediately addressed, please feel free to chime in.

**Comment by mickaelaccensi** [comment]
ww3_ounf.nml and ww3_shel.nml should also be updated to add this new output field

**Comment by MatthewMasarik-NOAA** [comment]
> ww3_ounf.nml and ww3_shel.nml should also be updated to add this new output field

Hi @mickaelaccensi
 
@JessicaMeixner-NOAA and I discussed this previously ([1034#discussion](https://github.com/NOAA-EMC/WW3/pull/1034#discussion_r1341483325)) with the outcome of deciding to limit the current PR to *.inp files, and had added issue #1111 for the expansion to *.nml.  Does that seem reasonable?

**Comment by mickaelaccensi** [comment]
ok, it seems acceptable to not expand this current PR with nml if it's done in another one already scheduled

**Comment by ukmo-ccbunney** [comment]
> @ukmo-ccbunney @mickaelaccensi @thesser1 - These updates will eventually be coming back to develop. If there's anything you see that needs to be immediately addressed, please feel free to chime in.
Nothing is jumping out at me, so all good as far as I am concerned. 👍

**Comment by MatthewMasarik-NOAA** [comment]
Thanks for your confirmations @mickaelaccensi @ukmo-ccbunney

**Comment by MatthewMasarik-NOAA** [comment]
@alperaltuntas @JessicaMeixner-NOAA 
I made small updates to the header, the main update checking off the list checkboxes.

File: model/src/w3iogomd.F90
Comment (DeniseWorthen) [inline model/src/w3iogomd.F90]: Please indent this code correctly (also LN 1657, 1662, 1681, 1964, 2011 below).

Comment (MatthewMasarik-NOAA) [reply model/src/w3iogomd.F90]: Sure, let me look into it.

Comment (MatthewMasarik-NOAA) [reply model/src/w3iogomd.F90]: I can't commit to this branch directly, but I'll get a branch in the works to raise a sub-pr.  The timeline is Monday.  @DeniseWorthen @JessicaMeixner-NOAA

Comment (MatthewMasarik-NOAA) [reply model/src/w3iogomd.F90]: @DeniseWorthen, I've addressed each of these lines in this PR to ESCOMP: https://github.com/ESCOMP/WW3/pull/20, which is now ready for review.

**Comment by zach1221** [comment]
Ok, testing is complete on WM #[2195](https://github.com/ufs-community/ufs-weather-model/pull/2195). @JessicaMeixner-NOAA can you please merge this PR for us?

**Comment by MatthewMasarik-NOAA** [comment]
A belated thank you and congratulations to @qingli411 and @alperaltuntas for this very nice contribution to WW3, which was merged into the dev/ufs-weather-model branch last Friday!  I'd also like to acknowledge @CarstenHansen, whose early posts to #669 provided helpful insight for implementing the namelist functionality.

**Pull request merged**
